### ISSUE TITLE
Feature: mixed encodings in a single line / custom encoding boundaries

### ISSUE THREAD

**Issue by fffergal** [LuminosoInsight/python-ftfy#51]
I'm looking into using ftfy to help upstream data sources clean up their data. a common format is CSV files. These data sources often manage to mix up encoding in different columns in the same row (maybe they're from different databases or web forms).

ftfy will be even more useful for me if it were possible to fix mixed encodings on the same line, or alternatively to add an argument to control what is considered a segment.

Obviously I could just make some code to use ftfy functions in CSV fields myself to get the desired results, but ftfy is a well documented easy to use tool so I'd really like offload the analysis to the bad data providers themselves.

If it's agreed this could be a feature, I might be able to spend some time on it.

**Comment by rspeer** [comment]
While I doubt that we could anticipate every kind of encoding boundary, an option to fix it by CSV fields sounds just as relevant as the default of fixing it by lines.

Applying `fix_text_segment` to every field read out of a CSV would also do a fine job here, of course. But I believe you're suggesting that ftfy should be usable as a way to fix a CSV file without writing explicit code that reads and rewrites the CSV file. So yes, if you make a working PR that makes that option available, I'd merge it.

Be sure to account for tab-separated files as well. The only CSV-like format you can get out of Excel that supports arbitrary Unicode at all, I believe, is tab-separated UTF-16, though presumably the UTF-16 should already be decoded by the time a function like this sees it.

**Comment by skyscanner-hpe** [comment]
You're right, this does sound like more of a CLI addition, it's already completely possible to do this easily in code.

Maybe I can expose the kwargs for csv.reader from a command to cover all the options.

What would you think about refactoring the CLI to use [click](http://click.pocoo.org/5/). It's the easiest way I know for opening files from command line (or stdin/stdout) as unicode in Python 2 and 3.

**Comment by Riprock** [comment]
You're right, this does sound like more of a CLI addition, it's already completely possible to do this easily in code.

Maybe I can expose the kwargs for csv.reader from a command to cover all the options.

What would you think about refactoring the CLI to use [click](http://click.pocoo.org/5/). It's the easiest way I know for opening files from command line (or stdin/stdout) as unicode in Python 2 and 3.

**Comment by rspeer** [comment]
Oh right, following up on this. I appreciate `click` and it looks like I could save a lot of code by using it now.

**Comment by rspeer** [comment]
...except, wait, the problem here is that ftfy takes the encoding as a command-line argument, but you have to have already told the encoding to `click` by the time it's parsing arguments. Sounds tricky.

**Comment by Riprock** [comment]
Yeah, that was the first thing I noticed when I started looking into it. Starting looking around at what other people had done. One option is closing the object that click gives you, changing the encoding then opening it again, but isn't exactly pretty. I'm not mad on using things like file without entering a context manager these days anyway. Maybe can propose and delayed open option to click...

Also hi! Yep, hit those snags a while ago and that's why I deferred it. That said I did add this to my personal backlog while I was on holiday, want to get back into my hobbies.
